src/librustdoc/html/render/print_item.rs

```diff
@@ -1869,107 +1914,122 @@ fn render_union<'a, 'cx: 'a>(
-    mut w: &mut Buffer,
-    it: &clean::Item,
-    g: Option<&clean::Generics>,
+fn render_struct<'a, 'cx: 'a>(
+    it: &'a clean::Item,
+    g: Option<&'a clean::Generics>,
     ty: Option<CtorKind>,
-    fields: &[clean::Item],
-    tab: &str,
+    fields: &'a [clean::Item],
+    tab: &'a str,
     structhead: bool,
-    cx: &Context<'_>,
-) {
-    let tcx = cx.tcx();
-    write!(
-        w,
-        "{}{}{}",
-        visibility_print_with_space(it.visibility(tcx), it.item_id, cx),
-        if structhead { "struct " } else { "" },
-        it.name.unwrap()
-    );
-    if let Some(g) = g {
-        write!(w, "{}", g.print(cx))
-    }
-    match ty {
-        None => {
-            let where_displayed =
-                g.map(|g| print_where_clause_and_check(w, g, cx)).unwrap_or(false);
+    cx: &'a Context<'cx>,
+) -> impl fmt::Display + 'a + Captures<'cx> {
+    display_fn(move |mut f| {
+        let tcx = cx.tcx();
+        write!(
+            f,
+            "{}{}{}",
+            visibility_print_with_space(it.visibility(tcx), it.item_id, cx),
+            if structhead { "struct " } else { "" },
+            it.name.unwrap()
+        )?;
 
-            // If there wasn't a `where` clause, we add a whitespace.
-            if !where_displayed {
-                w.write_str(" {");
-            } else {
-                w.write_str("{");
-            }
-            let count_fields =
-                fields.iter().filter(|f| matches!(*f.kind, clean::StructFieldItem(..))).count();
-            let has_visible_fields = count_fields > 0;
-            let toggle = should_hide_fields(count_fields);
-            if toggle {
-                toggle_open(&mut w, format_args!("{} fields", count_fields));
-            }
-            for field in fields {
-                if let clean::StructFieldItem(ref ty) = *field.kind {
-                    write!(
-                        w,
-                        "\n{}    {}{}: {},",
-                        tab,
-                        visibility_print_with_space(field.visibility(tcx), field.item_id, cx),
-                        field.name.unwrap(),
-                        ty.print(cx),
-                    );
-                }
-            }
+        if let Some(g) = g {
+            write!(f, "{}", g.print(cx))?;
+        }
 
-            if has_visible_fields {
-                if it.has_stripped_entries().unwrap() {
-                    write!(w, "\n{}    /* private fields */", tab);
+        match ty {
+            None => {
+                let mut buf = Buffer::html();
+                let where_displayed =
+                    g.map(|g| print_where_clause_and_check(&mut buf, g, cx)).unwrap_or(false);
+                write!(f, "{}", buf.into_inner())?;
+
+                // If there wasn't a `where` clause, we add a whitespace.
+                if !where_displayed {
+                    f.write_str(" {")?;
+                } else {
+                    f.write_str("{")?;
                 }
```

Comment:
We can merge this `write` with the next one:

```suggestion
                write!(f, "{}{}{{", buf.into_inner(), if !where_displayed { " " } else { "" })?;
```